assets/js/atomic/utils/blocks-registration-manager/blocks-registration-manager.ts

```diff
@@ -0,0 +1,181 @@
+	 *
+	 * @param {Object}  params                          - The parameters for the method.
+	 * @param {string}  params.blockWithRestrictionName - The name of the block with restrictions.
+	 * @param {string}  params.currentTemplateId        - The ID of the current template.
+	 * @param {boolean} params.isPostOrPage             - Whether the editor is in a post or page.
+	 * @return {boolean} True if the block should be registered, false otherwise.
+	 */
+	private blockShouldBeRegistered( {
+		blockWithRestrictionName,
+		currentTemplateId,
+		isPostOrPage,
+	}: {
+		blockWithRestrictionName: string;
+		currentTemplateId: string;
+		isPostOrPage: boolean;
+	} ) {
+		const {
+			allowedTemplates,
+			allowedTemplateParts,
+			availableInPostOrPageEditor,
+		} = this.blocksWithRestriction[ blockWithRestrictionName ];
+		const shouldBeAvailableOnTemplate = Object.keys(
+			allowedTemplates
+		).some( ( allowedTemplate ) =>
+			currentTemplateId.startsWith( allowedTemplate )
+		);
+		const shouldBeAvailableOnTemplatePart = Object.keys(
+			allowedTemplateParts
+		).some( ( allowedTemplate ) =>
+			currentTemplateId.startsWith( allowedTemplate )
+		);
+		const shouldBeAvailableOnPostOrPageEditor =
+			isPostOrPage && availableInPostOrPageEditor;
+
+		return (
+			shouldBeAvailableOnTemplate ||
+			shouldBeAvailableOnTemplatePart ||
+			shouldBeAvailableOnPostOrPageEditor
+		);
+	}
+
+	/**
+	 * Unregisters blocks before entering a restricted area based on the current template or page/post.
+	 *
+	 * This method iterates over all blocks with restrictions and unregisters them if they should not be registered
+	 * based on the current template ID and whether the editor is in a post or page. It uses a block registration
+	 * strategy to unregister the blocks, which depends on whether the block is a variation block or a regular block.
+	 *
+	 * @param {Object}  params                   - The parameters for the method.
+	 * @param {string}  params.currentTemplateId - The ID of the current template.
+	 * @param {boolean} params.isPostOrPage      - Whether the editor is in post or page mode.
+	 */
+	unregisterBlocksBeforeEnteringRestrictedArea( {
+		currentTemplateId,
+		isPostOrPage,
+	}: {
+		currentTemplateId: string;
+		isPostOrPage: boolean;
+	} ) {
+		for ( const blockWithRestrictionName of Object.keys(
+			BLOCKS_WITH_RESTRICTION
+		) ) {
+			if ( this.blocksWithRestriction[ blockWithRestrictionName ] ) {
+				if (
+					this.blockShouldBeRegistered( {
+						blockWithRestrictionName,
+						currentTemplateId,
+						isPostOrPage,
+					} )
+				) {
+					continue;
+				}
+				this.blockRegistrationStrategy = this.blocksWithRestriction[
+					blockWithRestrictionName
+				].isVariationBlock
+					? new BlockVariationStrategy()
+					: new BlockTypeStrategy();
+
+				this.blockRegistrationStrategy.unregister(
+					blockWithRestrictionName
+				);
+				unregisterBlockType( blockWithRestrictionName );
```

Comment:
I'm a bit confused by this, but ultimately these two instructions do the same, no? We are trying to unregister the block twice, causing a JS error:

![imatge](https://github.com/woocommerce/woocommerce-blocks/assets/3616980/3387636e-723b-40b7-9eca-b1d44229bc7d)

Reply:
This should be fixed in 6c809cf92ed85812eab292655b3699568054cafe.